subprojects/language-native/src/main/java/org/gradle/language/nativeplatform/internal/incremental/IncrementalNativeCompiler.java

```diff
@@ -54,15 +64,58 @@ public WorkResult execute(final T spec) {
         return workResult;
     }
 
+    private List<File> getSourceFilesForPch(T spec) {
+        // When the component defines a precompiled header, we need to check if the precompiled header is the _first_ header in the source file.
+        // For source files that do not include the precompiled header as the first file, we emit a warning
+        // For source files that do include the precompiled header, we mark them as a "source file for pch"
+        // The native compiler then adds the appropriate compiler arguments for those source files that can use PCH
+        if (spec.getPreCompiledHeader() != null) {
+            ImmutableList.Builder<File> sourceFiles = ImmutableList.builder();
+            for (File sourceFile : spec.getSourceFiles()) {
+                SourceFileState state = incrementalCompilation.getFinalState().getState(sourceFile);
+                final HashCode hash = state.getHash();
+                List<String> headers = Lists.newArrayList();
+                for (IncludeFileEdge edge : state.getEdges()) {
+                    if (hash.equals(edge.getIncludedBy())) {
```

Comment:
This is null for a `<>` include and so won't match. It would be better to ask the header parser for the details for the source file instead of reverse engineering this from the flattened include file graph, as this is what we actually need here. This way, we can change the incremental compile state without needed to rewrite this code here every time.